File: Top-Interview-Questions-Leetcode/Easy-Collections/other/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        prefix = []
        
        for ch in s:
            if ch == '(' or ch == '[' or ch == '{':
                prefix.append(ch)
            elif not prefix:
                return False
            elif abs(ord(prefix[-1]) - ord(ch)) > 2:
                return False
            else:
                prefix.pop()
        
        return not prefix

File: Top-Interview-Questions-Leetcode/Easy-Collections/other/valid-parentheses.cpp
#include <string>
#include <iostream>
#include <cmath>

class Solution {
public:
    bool isValid(std::string s) {
        int sz = s.size();
        std::string prefix = "";
        
        for (int i = 0; i < sz; ++i) {
            if (s[i] == '(' or s[i] == '{' or s[i] == '[') {
                prefix += s[i];
            }
            else if (prefix.size() == 0)
                return false;
            else if (abs(s[i] - prefix.back()) > 2) return false;
            else prefix.pop_back();
        }
        
        return prefix.size() == 0;
    }
};